[c10d][Fix] Start gloo sequence numbers at 0.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101422

Gloo PG used to create a random sequence number and broadcast it to
the rest of the group. But when we started enforcing sequence number checks in
ProcessGroupWrapper, we observed this was occasionally flaky. For example, this
error in a job was wrong, as all ranks were running the first broadcast
collective. Somehow the sequence number wasn't communicated across the store
correctly:

``
RuntimeError: Detected mismatch between collectives on ranks. Rank 16 is running collective: CollectiveFingerPrint(SequenceNumber=1977865401, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 1 is running collective: CollectiveFingerPrint(SequenceNumber=54090078, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))).Collectives differ in the following aspects: Sequence number: 1977865401vs 54090078
```

The issue reproduces rarely in tests, but is more common in large world size
jobs.

Differential Revision: [D45870688](https://our.internmc.facebook.com/intern/diff/D45870688/)

rohan-varma committed:
[c10d][Fix] Start gloo sequence numbers at 0.

Gloo PG used to create a random sequence number and broadcast it to
the rest of the group. But when we started enforcing sequence number checks in
ProcessGroupWrapper, we observed this was occasionally flaky. For example, this
error in a job was wrong, as all ranks were running the first broadcast
collective. Somehow the sequence number wasn't communicated across the store
correctly:

``
RuntimeError: Detected mismatch between collectives on ranks. Rank 16 is running collective: CollectiveFingerPrint(SequenceNumber=1977865401, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 1 is running collective: CollectiveFingerPrint(SequenceNumber=54090078, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))).Collectives differ in the following aspects: Sequence number: 1977865401vs 54090078
```

The issue reproduces rarely in tests, but is more common in large world size
jobs.

Differential Revision: [D45870688](https://our.internmc.facebook.com/intern/diff/D45870688/)

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2857,20 +2857,7 @@ void ProcessGroupGloo::monitoredBarrier(
-    uint64_t num = c10d::fromVec<char>(values);
-    sequenceNum_->set(num);
-  }
+  sequenceNum_ = c10d::SequenceNum(0);
```
does this still work if there are subgroups? Is sequence number only used for logging?

rohan-varma replied:
the only issue is a slightly worsened UX for debugging, because if multiple subgroups are there, it might be hard to tell which collective corresponds to which subgroup when there are issues in TDD=DETAIL mode.

rohan-varma committed:
Update on "[c10d][Fix] Start gloo sequence numbers at 0."


Gloo PG used to create a random sequence number and broadcast it to
the rest of the group. But when we started enforcing sequence number checks in
ProcessGroupWrapper, we observed this was occasionally flaky. For example, this
error in a job was wrong, as all ranks were running the first broadcast
collective. Somehow the sequence number wasn't communicated across the store
correctly:

``
RuntimeError: Detected mismatch between collectives on ranks. Rank 16 is running collective: CollectiveFingerPrint(SequenceNumber=1977865401, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 1 is running collective: CollectiveFingerPrint(SequenceNumber=54090078, OpType=BROADCAST, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))).Collectives differ in the following aspects: Sequence number: 1977865401vs 54090078
```

The issue reproduces rarely in tests, but is more common in large world size
jobs.

Differential Revision: [D45870688](https://our.internmc.facebook.com/intern/diff/D45870688/)

[ghstack-poisoned]

rohan-varma commented:
@XilunWu We probably need to root cause and fix the issue with the implementation of the store communicating the random number. Filed an issue:

rohan-varma commented:
@pytorchbot merge -f "CI done"
